pkg/controllers/machinesetmigration/machineset_migration_controller.go

```diff
@@ -0,0 +1,391 @@
+		// Set the authoritativeAPI status to Migrating.
+		if err := r.applyStatusAuthoritativeAPIWithPatch(ctx, mapiMachineSet, machinev1beta1.MachineAuthorityMigrating); err != nil {
+			return ctrl.Result{}, fmt.Errorf("unable to apply authoritativeAPI to status with patch: %w", err)
+		}
+
+		// Then wait for the old authoritativeAPI resource to have the Paused=True condition.
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Ensure the old authoritativeAPI resource has been requested to pause.
+	if err := r.ensurePausedRequestedOnOldAuthoritativeAPI(ctx, mapiMachineSet); err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to ensure old AuthoritativeAPI has been requested to pause: %w", err)
+	}
+
+	// Check if the old authoritativeAPI resource is actually paused.
+	if hasPausedConditionTrue, err := r.isOldAuthoritativeAPIStatusConditionPaused(ctx, mapiMachineSet); err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to check old AuthoritativeAPI has paused=true condition: %w", err)
+	} else if !hasPausedConditionTrue {
+		// The old Authoritative API resource is not paused yet, requeue to check later.
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Check if the synchronizedGeneration is up to date with the old authoritativeAPI's resource current generation.
+	if isSynchronizedGenMatching, err := r.isSynchronizedGenerationMatchingAuthoritativeAPIGeneration(ctx, mapiMachineSet); err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to check synchronizedGeneration is aligned: %w", err)
+	} else if !isSynchronizedGenMatching {
+		// The to-be Authoritative API resource is not fully synced up yet, requeue to check later.
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Check if the Synchronized condition is set to True.
+	// if it is not, this indicates an unmigratable resource and therefore we should error.
+	if getConditionStatus(mapiMachineSet, consts.SynchronizedCondition) != corev1.ConditionTrue {
+		// The Resource is NOT Synchronized, error.
+		return ctrl.Result{}, errors.New("unable to proceed migrating, resources are not synchronized")
+	}
+
+	// Add finalizer to new authoritative API.
+	// This will ensure no status changes on the same reconcile.
+	// The finalizer must be present on the object before we take any actions.
+	if addedFinalizer, err := r.ensureFinalizerOnNewAuthoritativeAPI(ctx, mapiMachineSet); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error adding finalizer: %w", err)
+	} else if addedFinalizer {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Remove finalizer from the old authoritative API.
+	// This will ensure no status changes on the same reconcile.
+	// The finalizer must be removed from the object before we take any actions.
+	if removedFinalizer, err := r.ensureFinalizerRemovedOnOldAuthoritativeAPI(ctx, mapiMachineSet); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error removing finalizer: %w", err)
+	} else if removedFinalizer {
+		return ctrl.Result{Requeue: true}, nil
+	}
```

Comment:
I wasn't expecting finalizer cleanup to be handled here, but in the sync controller deletion logic. I think having the finalizer on the resource here shouldn't be an issue AFAICT

Reply:
We discussed this offline, there are two kinds of finalizers, the CAPI/MAPI own ones and the new sync finalizer.
This is to ensure the first kind is set or removed accordingly, as described in https://issues.redhat.com/browse/OCPCLOUD-2564 migration controller behaviours
```
- Add appropriate Finalizer to new authoritative API, requeue here until we observe this in cache
- Remove Finalizer from old authoritative API, requeue here until we observe this in cache
```